### DISCUSSION TITLE
initialData-function that can return undefined changes the type of data after isSuccess check

### DISCUSSION THREAD

**Discussion by fredrik-andersson-yoin** [TanStack/query#5865] [Q&A] [answered]
Hi
I'm trying to update from v3 to v4 (tested v5 beta as well with the same issue) and having a problem that when using an initialData-function that can return undefined, we now get an error that data is possibly undefined after checking for isSuccess.

A function like this:
```
initialData: () => {
        const cachedTodo = queryClient.getQueryData<Todo[]>(['todos'])?.find((value) => value.id === id);
        return cachedTodo;
    }
```

Full example here:
https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgVwM4FMCKz1QJ4A0KG2ueAwgDbDoB28AvnAGZQQhwDkAAjAIa1U-AMYBrAPRR0fYTAC0ARxz5OAWABQG4PVzMZ6OABUIAEwiINcK3GAmAXHFrIQAI1wBuS9ZjoAHjAchKG0Ac091Bg0NYQhBeCUyKhp6OABeYixlCmo6GAAKAEpwmLibEzS4AEZi2KFEOBM+fiJgVABlZGFhdFRUOCZ0tEyyPIQvKwT8AGl0PAcAbU4YUwhOFpMAXQJxuEm8ADFaB0KHAAU2EFb0AB5jMwA+NMfz9iuAOilUCEoAN3RRsoOWxEHz+BycL4gAygmCcBgFbbqaw2WjAGDAPiUAAiTT4xwKTwsSOR1hKdWEMgAFugTHdzOk9klcm8QugYKR8Dj+LcVvMNvc8otlmZUJwNgUAPxvZjaEx5PI-TE4AmpR6Kyg4N62NKpdK2Io7ZFSGDIKC0OAU4TU2krcLIyIRA2adTAZhwPKtDpdHqoAljYlWMnfdBvSgQEJ5Rr8N4wopwcTiOAAUSgbCgcGpUjgbgpQzgEDdeYLKLRGOxuLeydT0AcUb4Nj6kF6wBclDwKFoJnQMtoNKiAfjicM1LguDTDVafDA0igfQA7tTzVaBCFQnAYCPQJBYOvzG99ywLlwpDJ5HtOHA5HI1QAWcQ-ADME+YzFwdG6bw0Do0CbgADkIHgKBkFoWg+FbdAWkoDUgiaNcx2gfNaDbIA

Is there a way to get around this, other than checking data for undefined everywhere?

**Comment by TkDodo** [comment, marked as answer]
interesting. seems like our generic gets `undefined` added into `TData`, as the return type of the Query after checking for is success is now:

```
QueryObserverSuccessResult<Todo | undefined, DefaultError>.data: Todo | undefined
```

This seems to be because of our overloads, and it's the same for v5. Probably worth opening an issue 👍

**Reply by TkDodo** [reply]
here's a failing test case btw, definitely a types bug:

```
it('TData should be narrowed after an isSuccess check when initialData is provided as a function which can return undefined', () => {
  doNotExecute(() => {
    const { data, isSuccess } = useQuery({
      queryKey: ['key'],
      queryFn: () => {
        return {
          wow: true,
        }
      },
      initialData: () => undefined as { wow: boolean } | undefined,
    })

    if (isSuccess) {
      const result: Expect<Equal<{ wow: boolean }, typeof data>> = true
      return result
    }
    return false
  })
})
```

**Reply by fredrik-andersson-yoin** [reply]
I'll open an issue then.

**Reply by TkDodo** [reply]
i have a fix already:
- https://github.com/TanStack/query/pull/5869

**Reply by fredrik-andersson-yoin** [reply]
Awesome!

**Reply by TkDodo** [reply]
should already be available on latest v5 beta. enjoy 🎉
